### PR TITLE
Native deploy implementation

### PR DESCRIPTION
**Pull request by pradithya**
**What this PR does / why we need it**:

Add ability for user to specify the deployment mode:
1. serverless (default): using knative
2. raw_deployment: using k8s deployment
This translate to annotation in InferenceService created by templater.go .

NOTE: changing deployment mode of an existing deployed model is possible only when the model is not in service state. This is due to the fact that it can happen gracefully (there is downtime)

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
Add deployment mode configuration
```

**Checklist**

- [x] Added unit test, integration, and/or e2e tests
- [ ] Tested locally
- [ ] Updated documentation
- [x] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

File: api/api/version_endpoints_api.go [hunk 1]
```diff
@@ -251,6 +256,11 @@ func (c *EndpointsController) UpdateEndpoint(r *http.Request, vars map[string]st
 			}
 		}
 
+		// Should not allow redeploying a "serving" model and changing its deployment type since we can't guaratee graceful deployment
+		if endpoint.IsServing() && newEndpoint.DeploymentType != endpoint.DeploymentType {
```
Comment (ariefrahmansyah) [inline api/api/version_endpoints_api.go:260]: So we allow the redeployment if:
1. Endpoint status is not serving so we can change the deployment type
2. Endpoint status is serving and we do not change the deployment type

File: api/api/version_endpoints_api_test.go [hunk 2]
```diff
@@ -3617,6 +3617,121 @@ func TestUpdateEndpoint(t *testing.T) {
 				data: Error{Message: "custom predictor image must be set"},
 			},
 		},
+		{
```
Comment (ariefrahmansyah) [inline api/api/version_endpoints_api_test.go:3620]: Can you please add one more test where the deployment status is running and it change the deployment type?

File: api/models/version_endpoint.go [hunk 3]
```diff
@@ -43,10 +43,18 @@ type VersionEndpoint struct {
...
 
 	CreatedUpdated
 }
 
+type DeploymentMode string
+
+const (
+	ServerlessDeploymentMode = "serverless"
+	RawDeploymentMode        = "raw_deployment"
+)
```
Comment (ariefrahmansyah) [inline api/models/version_endpoint.go:56]: nitpick:

```
const (
	ServerlessDeploymentMode DeploymentMode = "serverless"
	RawDeploymentMode DeploymentMode        = "raw_deployment"
)
```

**Review by ariefrahmansyah — APPROVED** [review]
lgtm with nit comments.

sdk and ui changes will be on different PR(s) right?

File: api/api/version_endpoints_api.go (hunk 1 above)
Comment (pradithya) [reply api/api/version_endpoints_api.go:260]: Yes

File: api/api/version_endpoints_api_test.go (hunk 2 above)
Comment (pradithya) [reply api/api/version_endpoints_api_test.go:3620]: Done

**Comment by pradithya** [comment]
> lgtm with nit comments.
> 
> sdk and ui changes will be on different PR(s) right?

Yes, SDK and UI will be in different UI

**Pull request merged**